### DISCUSSION TITLE
Create a simple to use wrapper on top of OpenSSL

### DISCUSSION THREAD

**Discussion by medvednikov** [vlang/v#11581] [Ideas]

**Comment by joe-conigliaro** [comment]
I think down the track we should use SChannel for windows and OpenSSL for nix. What do you think? Would be good not to have dependency on OpenSSL when on windows.

**Comment by medvednikov** [comment]
Yes, ideally.

But it's incredibly complex. And the documentation is terrible. I personally wouldn't even know how to approach it.

If there's someone with SChannel experience, who could help, that would be great.

**Comment by joe-conigliaro** [comment]
Doesn't GO have their own implementation? I think they don't use OpenSSL or SChannel, if I remember correctly they wrote it from scratch (I may be wrong). If they did, maybe that could be ported to V?

**Comment by medvednikov** [comment]
They have their own implementation, it uses a lot of Go assembly, so it's not portable.

It's also slower and less tested than OpenSSL.

**Comment by joe-conigliaro** [comment]
We should take example from here: https://github.com/sfackler/rust-native-tls

**Comment by Bowero** [comment]
OpenSSL is an option?: https://github.com/vlang/v/issues/307

**Comment by watzon** [comment]
For anyone familiar with the Ruby/Crystal syntax [this](https://github.com/crystal-lang/crystal/blob/c9d1eef8fde5c7a03a029d64c8483ed7b4f2fe86/src/openssl.cr) could be a good reference for creating OpenSSL bindings. I may take a crack at this myself.

**Comment by joe-conigliaro** [comment]
It would be really nice if at some stage we didn't have openssl as a dependency on win.
- windows: Secure Channel
- osx: Security Framework
- linux/others: openssl

I know it would be a lot of tedious work, but it would be nice

**Comment by medvednikov** [comment]
Absolutely, I'm all for having as few dependencies as possible.

**Comment by medvednikov** [comment]
This was implemented way back by @joe-conigliaro

**Comment by medvednikov** [comment]
Sorry I misread the issue, the openssl wrapper is still not implemented.
